Allow focusing into cross-origin iframes that have never had user interaction if currently processing a user interaction

beidson opened the pull request:
#### a513eb610e7a089a5c5b2d1904ac661a1f68ede9
<pre>
Allow focusing into cross-origin iframes that have never had user interaction if currently processing a user interaction
<a href="https://rdar.apple.com/153007151">rdar://153007151</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=294296">https://bugs.webkit.org/show_bug.cgi?id=294296</a>

Reviewed by Ryosuke Niwa and Alex Christensen.

When tabbing through a document, you might reach the end of the current frame&apos;s focusable elements
and dig into an iframe to find the first focusable one there.

When we do this today, if the iframe is cross-origin and has never been interacted with, we find
the element but refuse to actually focus it.

This makes sense to prevent programatic JS focus into a cross-origin iframe, but does not make sense
when the user is driving the focusing themselves (e.g. pressing tab)

Making this change matches other browsers.

* LayoutTests/http/tests/dom/focus-in-cross-origin-iframe-expected.txt: Added.
* LayoutTests/http/tests/dom/focus-in-cross-origin-iframe.html: Added.
* LayoutTests/http/tests/dom/resources/focus-in-cross-origin-iframe-iframe.html: Added.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::focus): Add an &quot;is NOT processing user gesture&quot; qualification to the early return for
  focus into a cross-origin iframe.

Canonical link: <a href="https://commits.webkit.org/296110@main">https://commits.webkit.org/296110@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c23df231c40ded80641dec02c4f4a1693d0ee533

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107391 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27073 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17477 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112605 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/57927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109355 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27755 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35573 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81522 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/57927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110320 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21987 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96792 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61897 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21427 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14923 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57371 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91356 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14957 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115706 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34457 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/25396 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90563 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34833 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93041 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90300 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23027 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35210 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12999 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30201 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34379 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/39920 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34125 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37480 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35786 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 reviewed (changes requested):
Test says FAILED

rniwa commented on LayoutTests/http/tests/dom/focus-in-cross-origin-iframe.html:
```diff
@@ -0,0 +1,33 @@
+}
+
+window.addEventListener("message", (event) => {
+	document.getElementById("result").innerText = event.data;
```
Tab.

achristensen07 replied:
I think it's especially appropriate to have tabs in this test

rniwa commented on LayoutTests/http/tests/dom/focus-in-cross-origin-iframe.html:
```diff
@@ -0,0 +1,33 @@
+
+function startTest()
+{
+	if (!eventSender)
```
Tab.

rniwa commented on LayoutTests/http/tests/dom/resources/focus-in-cross-origin-iframe-iframe.html:
```diff
@@ -0,0 +1,18 @@
+}
+
+window.addEventListener("message", (event) => {
+	if (document.activeElement == document.getElementById("targetElement"))
```
Tab

beidson replied:
Alex encouraged me to leave the tabs in - more variety of input in tests is better 😂

rniwa commented on LayoutTests/http/tests/dom/focus-in-cross-origin-iframe.html:
```diff
@@ -0,0 +1,33 @@
+	if (!eventSender)
+		return
+		
+	eventSender.keyDown("\t")
```
Can we use `UIHelper.keyDown` instead (after including `LayoutTests/resources/ui-helper.js`)? That'll make this test also work in iOS.

beidson replied:
Never heard of it. Just assumed iOS focus worked differently (because... it does)

I'll try that though.

beidson replied:
Yup, iOS focus is still different enough for it to need an exception here. That's fine.

beidson committed:
Allow focusing into cross-origin iframes that have never had user interaction if currently processing a user interaction
rdar://153007151
https://bugs.webkit.org/show_bug.cgi?id=294296

Reviewed by Ryosuke Niwa and Alex Christensen.

When tabbing through a document, you might reach the end of the current frame's focusable elements
and dig into an iframe to find the first focusable one there.

When we do this today, if the iframe is cross-origin and has never been interacted with, we find
the element but refuse to actually focus it.

This makes sense to prevent programatic JS focus into a cross-origin iframe, but does not make sense
when the user is driving the focusing themselves (e.g. pressing tab)

Making this change matches other browsers.

* LayoutTests/http/tests/dom/focus-in-cross-origin-iframe-expected.txt: Added.
* LayoutTests/http/tests/dom/focus-in-cross-origin-iframe.html: Added.
* LayoutTests/http/tests/dom/resources/focus-in-cross-origin-iframe-iframe.html: Added.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::focus): Add an "is NOT processing user gesture" qualification to the early return for
  focus into a cross-origin iframe.

Canonical link: https://commits.webkit.org/296110@main
